src/odemis/driver/smaract.py

```diff
@@ -3173,3 +2961,1175 @@ def SA_CTL_Move(self, handle, ch, pos_pm, _):
+                if future._must_stop:
+                    future._was_stopped = True
+                # Switch to automatic adjustment
+                self.SetProperty_i32(SA_PS_AF_ADJUSTMENT_STATE_PROP,
+                                     SA_PS_ADJUSTMENT_STATE_AUTO_ADJUST)
+            self._wait_for_event(SA_PS_ADJUSTMENT_STATE_DISABLED)  # state will be DISABLED when done
+            logging.debug("Finished referencing.")
+
+            # We could save the referencing parameters to memory here. However, it is only possible
+            # to save these parameters together with the configuration parameters (fiber length etc.).
+            # In case something goes wrong, there is a chance that the device needs to be reconfigured.
+            # Since referencing should be done on startup anyway, there is no real need to save the
+            # parameters.
+        except SA_SIError as ex:
+            # This occurs if a stop command interrupts referencing
+            if ex.errno == SA_SI_ERROR_CANCELLED:
+                future._was_stopped = True
+                logging.info("Referencing stopped: %s", ex)
+                raise CancelledError()
+            elif future._must_stop:
+                future._was_stopped = True
+                raise CancelledError()
+            else:
+                logging.error("Referencing failed: %s", ex)
+                raise
+        except CancelledError:
+            logging.debug("Referencing cancelled.")
+            future._was_stopped = True
+            raise  # No fuss, pass it as-is
+        except Exception:
+            logging.exception("Referencing failure.")
+            raise
+        finally:
+            # If the state of the device is stable and all channels are valid, the referencing procedure
+            # succeeded. Typically, all channels become valid at the same time during the reference procedure.
+            # We still check all of them inidividually to be sure.
+            if self.IsStable():
+                for ch, num in self._channels.items():
+                    self.referenced._value[ch] = self.IsValid(num)
+            self._updatePosition()
+            # We only notify after updating the position so that when a listener
+            # receives updates both values are already updated.
+            self.referenced.notify(self.referenced.value)
+
+            # The adjustment state should be "disabled" after automatic adjustment. If the referencing
+            # procedure was stopped prematurely, right after it was set to "manual", it might be in the
+            # wrong state --> make sure it's set to disabled here.
+            if self.GetAdjustmentState() != SA_PS_ADJUSTMENT_STATE_DISABLED:
+                self.SetProperty_i32(SA_PS_AF_ADJUSTMENT_STATE_PROP,
+                                     SA_PS_ADJUSTMENT_STATE_DISABLED)
+                self._wait_for_event(SA_PS_ADJUSTMENT_STATE_DISABLED)  # state will be DISABLED when done
```

Comment:
For safety, I'd suggest to always put some timeout. It can be large values (eg, 10s, 5 min)... but they end up always much smaller than infinity.